File: 2024/03/202403.py
import re
input = open("input.txt").read()
muls = re.findall(r'mul\(\d+,\d+\)', input)
pairs = [[int(n) for n in re.findall(r'\d+', m)] for m in muls]
print(sum(x*y for x,y in pairs))

insts = re.findall(r"mul[(]\d+,\d+[)]|do[(][)]|don't[(][)]", input)
r = 0
en = 1
for inst in insts:
  if inst.startswith("mul"):
    if en:
      a,b = [int(x) for x in inst[4:-1].split(",")]
      r += a * b
  elif inst.startswith("don't"):
    en = 0
  elif inst.startswith("do"):
    en = 1

print(r)

File: 2024/03/202403.lua
require 'pl'
stringx.import()
pp = pretty.dump
inp = file.read("input.txt")

function main(part)
  s = 0
  m = 1
  for i = 1,#inp do
    if inp:find("^don't%(%)", i) then
      m = (part == 1) and 1 or 0
    elseif  inp:find("^do%(%)", i) then
      m = 1
    else
      x,y,a,b = inp:find("^mul%((%d+),(%d+)%)", i)
      if x then
        s = s + m * ((a+0)<1000 and a or 0) * ((b+0)<1000 and b or 0)
      end
    end
  end
  return s
end

print(main(1))
print(main(2))
